heron/schedulers/src/java/com/twitter/heron/scheduler/mesos/framework/MesosFramework.java

```diff
@@ -0,0 +1,575 @@
+  protected void handleMesosStatusUpdate(String taskId,
+                                         Protos.TaskStatus status) {
+    // We would determine what to do basing on current taskId and mesos status update
+    switch (status.getState()) {
+      case TASK_STAGING:
+        LOG.info(String.format("Task with id '%s' STAGING", status.getTaskId().getValue()));
+        LOG.severe("Framework status updates should not use");
+
+        break;
+      case TASK_STARTING:
+        LOG.info(String.format("Task with id '%s' STARTING", status.getTaskId().getValue()));
+
+        break;
+      case TASK_RUNNING:
+        LOG.info(String.format("Task with id '%s' RUNNING", status.getTaskId().getValue()));
+
+        break;
+      case TASK_FINISHED:
+        LOG.info(String.format("Task with id '%s' FINISHED", status.getTaskId().getValue()));
+        LOG.severe("Heron job should not finish!");
+
+        break;
+      case TASK_FAILED:
+        LOG.info(String.format("Task with id '%s' FAILED", status.getTaskId().getValue()));
+        handleMesosFailure(taskId);
+
+        break;
+      case TASK_KILLED:
+        LOG.info("Task killed which is supported running. "
+            + "It could be triggered by manual restart command");
+        handleMesosFailure(taskId);
+
+        break;
+      case TASK_LOST:
+        LOG.info(String.format("Task with id '%s' LOST", status.getTaskId().getValue()));
+        handleMesosFailure(taskId);
+
+        break;
+      case TASK_ERROR:
+        LOG.info(String.format("Task with id '%s' ERROR", status.getTaskId().getValue()));
+        handleMesosFailure(taskId);
+
+        break;
+      default:
+        LOG.severe("Unknown TaskState:" + status.getState() + " for taskId: "
+            + status.getTaskId().getValue());
+
+        break;
+    }
+  }
+
+
+  /**
+   * Restart a failed task unless exceeding the retires limitation
+   *
+   * @param taskId the taskId of container to handle
+   */
+  protected void handleMesosFailure(String taskId) {
+    int attempt = TaskUtils.getAttemptForTaskId(taskId);
+    BaseContainer container = containersInfo.get(TaskUtils.getContainerIndexForTaskId(taskId));
+
+    boolean hasAttemptsLeft = attempt < container.retries;
+
+    if (hasAttemptsLeft) {
+      LOG.warning(String.format("Retrying task: %s, attempt: %d", container.name, attempt + 1));
+
+      String newTaskId = TaskUtils.getTaskId(container.name, attempt + 1);
+      scheduleNewTask(newTaskId);
+    } else {
+      LOG.severe("Would not restart the job since it is beyond retries: " + attempt);
+    }
+  }
+
+  /**
+   * Schedule a new task
+   *
+   * @param taskId the taskId of container to handle
+   */
+  protected boolean scheduleNewTask(String taskId) {
+    // Put it to the pending start queue
+    LOG.info(String.format("We are to schedule task: [%s]", taskId));
+
+    // Update the tasksId
+    int containerIndex = TaskUtils.getContainerIndexForTaskId(taskId);
+    tasksId.put(containerIndex, taskId);
```

Comment:
handleMesosFailure is a protected method invoked in a public method, which is within the synchronized block; so `handleMesosFailure` and `killJob` will not concurrently executed.
It may not be a good idea to add synchronized block in an internal method, i.e. private or protected.